src/_P103_Atlas_EZO_pH_ORP_EC.ino

```diff
@@ -34,496 +36,528 @@ boolean Plugin_103(byte function, struct EventStruct *event, String& string)
-            }
-            break;
-          }
+        cmd += F("mid,");
+        cmd += PCONFIG_FLOAT(1);
+        triggerCalibrate = true;
       }
-
-      if (triggerCalibrate) {
-        char sensordata[32];
-        _P103_send_I2C_command(I2Cchoice, cmd, sensordata);
+      else if (isFormItemChecked(F("Plugin_103_enable_cal_L")))
+      {
+        cmd += F("low,");
+        cmd += PCONFIG_FLOAT(2);
+        triggerCalibrate = true;
+      }
+      else if (isFormItemChecked(F("Plugin_103_enable_cal_H")))
+      {
+        cmd += F("high,");
+        cmd += PCONFIG_FLOAT(3);
+        triggerCalibrate = true;
       }
+      break;
+    }
 
-      if(board_type == PH || board_type == EC) {
-        char   deviceTemperatureTemplate[40];
-        String tmpString = web_server.arg(F("Plugin_103_temperature_template"));
-        strncpy(deviceTemperatureTemplate, tmpString.c_str(), sizeof(deviceTemperatureTemplate) - 1);
-        deviceTemperatureTemplate[sizeof(deviceTemperatureTemplate) - 1] = 0; // be sure that our string ends with a \0
+    case ORP:
+    {
+      PCONFIG_FLOAT(1) = getFormItemFloat(F("Plugin_103_ref_cal_O"));
 
-        addHtmlError(SaveCustomTaskSettings(event->TaskIndex, (byte *)&deviceTemperatureTemplate, sizeof(deviceTemperatureTemplate)));
+      if (isFormItemChecked(F("Plugin_103_enable_cal_O")))
+      {
+        cmd += PCONFIG_FLOAT(1);
+        triggerCalibrate = true;
       }
-
-      success = true;
       break;
     }
 
-    case PLUGIN_INIT:
+    case EC:
     {
+      PCONFIG(3) = getFormItemInt(F("Plugin_103_ref_cal_single"));
+      PCONFIG(4) = getFormItemInt(F("Plugin_103_ref_cal_L"));
+      PCONFIG(5) = getFormItemInt(F("Plugin_103_ref_cal_H"));
+
+      if (isFormItemChecked(F("Plugin_103_enable_cal_dry")))
+      {
+        cmd += F("dry");
+        triggerCalibrate = true;
+      }
+      else if (isFormItemChecked(F("Plugin_103_enable_cal_single")))
+      {
+        cmd += PCONFIG(3);
+        triggerCalibrate = true;
+      }
+      else if (isFormItemChecked(F("Plugin_103_enable_cal_L")))
+      {
+        cmd += F("low,");
+        cmd += PCONFIG(4);
+        triggerCalibrate = true;
+      }
+      else if (isFormItemChecked(F("Plugin_103_enable_cal_H")))
+      {
+        cmd += F("high,");
+        cmd += PCONFIG(5);
+        triggerCalibrate = true;
+      }
       break;
     }
+    }
 
-    case PLUGIN_READ:
+    if (triggerCalibrate)
     {
-      board_type = PCONFIG(0);
-      I2Cchoice  = PCONFIG(1);
+      char calibration[ATLAS_EZO_RETURN_ARRAY_SIZE] = {0};
+      _P103_send_I2C_command(I2Cchoice, cmd, calibration);
+    }
 
-      char sensordata[32];
-      bool status;
-      String readCommand;
+    if (board_type == PH || board_type == EC)
+    {
+      char deviceTemperatureTemplate[40] = {0};
+      String tmpString = web_server.arg(F("Plugin_103_temperature_template"));
+      strncpy(deviceTemperatureTemplate, tmpString.c_str(), sizeof(deviceTemperatureTemplate) - 1);
```

Comment:
there is also a `safe_strncpy` function.

Reply:
fixed